File: 1/2020.09.25/剑指Offer 40.最小的k个数/java/solution.java
class Solution {
    public int[] getLeastNumbers(int[] arr, int k) {
        if (k == 0) {
            return new int[0];
        }
        PriorityQueue<Integer> heap = new PriorityQueue(k, new Comparator<Integer>() {
            @Override
            public int compare(Integer o1, Integer o2) {
                return o2 - o1;
            }
        });
        for (int i = 0; i < k; i++) {
            heap.add(arr[i]);
        }
        for (int i = k; i < arr.length; i++) {
            if (arr[i] < heap.peek()) {
                heap.poll();
                heap.add(arr[i]);
            }
        }
        int[] res = new int[k];
        Iterator<Integer> it = heap.iterator();
        for (int i = 0; i < res.length; i++) {
            res[i] = it.next();
        }
        return res;
    }
}

File: 1/2020.09.25/剑指Offer 40.最小的k个数/python/solution.py
import heapq
class Solution:
    def getLeastNumbers(self, arr: List[int], k: int) -> List[int]:
        if k == 0:
            return []
        heap = []
        for i in range(k):
            heapq.heappush(heap, -arr[i])
        for i in range(k, len(arr)):
            if -arr[i] > heap[0]:
                heapq.heappop(heap)
                heapq.heappush(heap, -arr[i])
        return [-i for i in heap]

    def getLeastNumbers(self, arr: List[int], k: int) -> List[int]:
        return sorted(arr)[:k]
